[reliability][enhance]reliability cerberus integration

qiliRedHat opened the pull request:
https://issues.redhat.com/browse/OCPQE-4515
[Reliability Test Tool Enhancement Plan ](https://docs.google.com/document/d/12n_yOXxZcsaokEE4rt04qpj3z7p9eWiSk4m5UqTEY7U/edit#heading=h.n4ceby94lm77)

**What**
Leverage [Cerberus](https://github.com/cloud-bulldozer/cerberus) to monitor the health of the cluster and provide go/no-go signal to reliability test to make go/no-go decisions before starting next task.

**Why**
Current Reliability test needs to check periodically manually for the health of the cluster. This enhancement reduces manual monitoring effort, and together with [Cerberus's slack notification](https://github.com/cloud-bulldozer/cerberus/blob/master/docs/slack.md) feature, reliability test owner can ack problem in time.

**How**
Reliability test checks the Cerberus reported cluster status via http://0.0.0.0:8080. If Cerberus reported 'False' status, according to the configured of cerberus_fail_action, reliability test will
1. pause reliability test and allow for test owner to resolve the failure reported by Cerberus and continue.
2. halt reliability test.
3. continue reliability test but warn in log about the Cerberus 'False' result.
example of the config file piece:
```
  cerberusIntegration:
    # start cerberus https://github.com/cloud-bulldozer/cerberus before starting reliabiity test.
    cerberus_enable: True
    # if cerberus_enable is false, the following 2 items are ignored.
    cerberus_api: "http://0.0.0.0:8080"
    # action to take when cerberus status is False, valid data: pause/halt/continue
    cerberus_fail_action: pause
```

**Exit Criteria**
Reliability test can eanble Cerberus integration and configure the action to take when Cerberus reported status is 'False' (cluster is unhealthy).

qiliRedHat commented:
PTAL @mffiedler @paigerube14 @kedark3 @wabouhamad

kedark3 commented on reliability/config/enhance_reliability.yaml:
```diff
@@ -12,6 +12,13 @@ reliability:
     maxProjects: 25
     sleepTime: 10
+  cerberusIntegration:
+    # start cerberus https://github.com/cloud-bulldozer/cerberus before starting reliabiity test.
```
While GitHub Repo for Cerberus remains the official source of starting the cerberus, here's a tiny script we can link to here in the repo to make it easier: 
```run_cerberus.sh 
#!/bin/sh
function getKubeconfigFromFlexy(){
curl --insecure https://<jenkins_url>/job/ocp-common/job/Flexy-install/$BUILD_ID/artifact/workdir/install-dir/auth/kubeconfig -o ~/.kube/config$BUILD_ID ; export KUBECONFIG=~/.kube/config$BUILD_ID


}

echo "Invocation as run_cerberus.sh <FlexybuildId> <hostportmapping>"
BUILD_ID=$1
PORT=$2
echo "Using build $BUILD_ID and port $PORT"
set -x
# getKubeconfigFromFlexy

sudo firewall-cmd --add-port=$PORT/tcp --permanent
sudo firewall-cmd --reload

podman logs -f `podman run --name=cerberus$BUILD_ID -e SLACK_API_TOKEN="<token>" -e SLACK_CHANNEL=G01195FMYLF -p $PORT:8080 -v ~/.kube/config$BUILD_ID:/root/.kube/config:Z -v ~/cerberus_config.yaml:/root/cerberus/config/config.yaml:Z -d quay.io/openshift-scale/cerberus:latest`
```


This will create and start tailing the logs for a cerberus container. Also expose ports through firewall.

kedark3 commented on reliability/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+pyyaml
```
I think it would be a good idea to lock the version for each of these - as unwanted updates can break things. Something we always followed in our previous team.

qiliRedHat replied:
@kedark3 Good point! Thanks for sharing your experience.

qiliRedHat replied:
@kedark3 PTAL https://github.com/openshift/svt/pull/677/commits/65b1fd678d96caf724ef577c41c238020561c716

kedark3 replied:
that looks good @qiliRedHat

chaitanyaenr commented on reliability/config/enhance_reliability.yaml:
```diff
@@ -12,6 +12,13 @@ reliability:
     # for 5 nodes m5.xlarge cluster, 60 is recomended
     maxProjects: 25
     sleepTime: 10
+  cerberusIntegration:
```
@qiliRedHat This is a great addition. Cebrerus stores all the failures in a database and we can hit the metrics API to get the history of the failures for the run which in this is the reliability test run duration: https://github.com/cloud-bulldozer/cerberus#metrics-api. It might be useful to save it as an artifact in addition to the logs. Thoughts?

paigerube14 replied:
Maybe if the status is false we could do a second call specifically to the history api to print out the issues cerberus encountered

qiliRedHat replied:
@chaitanyaenr @paigerube14 Thank you for your comments! I added a function to save the Cerberus history whenever there is new failure items added to the history, if there is no new failure item added, nothing will be done. No mater the Cerberus status check result is `True` or `False` this save function is called always. The reason of doing this instead of saving history when Cerberus status is `False` is, Reliability test checks the Cerberus status in a certain interval, some times the `False` status is recovered to `True` between 2 checks, in this situation, Reliability gets `True` status, but history shows some failure.
The file where the Cerberus history is saved can be configured  or default to /tmp/cerberus-history.json.
PTAL https://github.com/openshift/svt/pull/677/commits/b7ab565fc4e2747ba103b1ce2cca7f7037a268bb

Test I run:
Kill one pod
```
 % oc delete pod node-exporter-nns7m -n openshift-monitoring
pod "node-exporter-nns7m" deleted
```
Cerberus did not hit the `False` status in its check,  but history of the failure is saved
```
2021-07-12 16:22:15,576 - INFO - Cerberus status is 'True'.
2021-07-12 16:22:15,583 - INFO - Cerberus history is updated. Latest history is saved to './cerberus-history.json'
```
```
 cat cerberus-history.json
{
    "history": {
        "failures": [
            {
                "timestamp": "2021-07-12 16:22:11",
                "count": 1,
                "issue": "pod crash",
                "name": "node-exporter-sh8kr",
                "component": "monitoring"
            }
        ]
    }
}%  
```
Kill another pod
```
% oc delete pod node-exporter-5tqm8 -n openshift-monitoring
pod "node-exporter-5tqm8" deleted
```
Cerberus did not hit the `False` status in its check,  but history of the failure is saved
```
2021-07-12 16:23:28,418 - INFO - Cerberus status is 'True'.
2021-07-12 16:23:28,424 - INFO - Cerberus history is updated. Latest history is saved to './cerberus-history.json'

cat cerberus-history.json
{
    "history": {
        "failures": [
            {
                "timestamp": "2021-07-12 16:22:11",
                "count": 1,
                "issue": "pod crash",
                "name": "node-exporter-sh8kr",
                "component": "monitoring"
            },
            {
                "timestamp": "2021-07-12 16:23:24",
                "count": 1,
                "issue": "pod crash",
                "name": "node-exporter-mkptx",
                "component": "monitoring"
            }
        ]
    }
}%   
```

paigerube14 commented:
This doesn't have to do with this specific PR but on line https://github.com/qiliRedHat/svt/blob/92c4e7a519c3651ab741ef0dca4f6117169aca5e/reliability/tasks/utils/oc.py#L16 are we able to print this output only when rc != 0. 
When doing the monitor cluster operators, it is a lot of extra output that is not useful if all operators are healthy. If the operator is degraded we are already printing an error https://github.com/qiliRedHat/svt/blob/92c4e7a519c3651ab741ef0dca4f6117169aca5e/reliability/tasks/Monitor.py#L22

paigerube14 commented:
Since we are adding in a requirements file in this PR, are we able to add the installation of the requirements to the README. Like: https://github.com/cloud-bulldozer/cerberus/blob/master/docs/installation.md#install-the-dependencies

qiliRedHat commented:
@paigerube14 @mffiedler I updated the README.md to include updated configurations and dependencies install. PTAL https://github.com/openshift/svt/pull/677/commits/b026b5b6370212e5728fd0bc6ea6d7c4ecd4a5cb

paigerube14 commented on reliability/README.md:
```diff
@@ -1,81 +1,205 @@
+
+If you installed your cluster with [Flexy-install](https://mastern-jenkins-csb-openshift-qe.apps.ocp4.prod.psi.redhat.com/job/ocp-common/job/Flexy-install/), download kubeconfig, users.spec and kubeadmin-password files. If you installed cluster in other way, prepare the above files accordingly.
+
+Replace kubeconfig kubeadmin_file and user_file in your config file.
```
Think it should be 
`kubeconfig, kubeadmin_password and user_file`

qiliRedHat replied:
I mean these 3 files in the screenshot below. 
<img width="503" alt="Screen Shot 2021-07-13 at 11 06 28 AM" src="https://user-images.githubusercontent.com/81630527/125384244-34490900-e3cb-11eb-9b06-71c7b00bb786.png">

Why do you say kubeconfig, kubeadmin_password and user_file? Where do you see them or do you just mean to use amore common names?

paigerube14 replied:
Just to have a more common names. What you have now "kubeadmin_file" is a little confusing and I wasn't 100% sure what you meant. I think changing "kubeadmin_file" to kubeadmin-password would make it a little more clear.

qiliRedHat replied:
make sense, I'll update the config file.

paigerube14 commented on reliability/README.md:
````diff
@@ -1,81 +1,205 @@
+    - user_file: <path_to_users.spec>
+```
+
+### Tasks timeSubstitutions
````
add space between time and substitutions

paigerube14 commented on reliability/README.md:
````diff
@@ -1,81 +1,205 @@
+```
+
+### Limits
+Limits limits the max project can be created during the reliability test. The number depends on the size of the cluster. Some recomendations can be found in the example [config](https://github.com/openshift/svt/blob/master/reliability/config/example_reliability.yaml)
````
think it should be `Limits limit the max project**s that** can be created...`

paigerube14 commented on reliability/README.md:
```diff
@@ -1,81 +1,205 @@
+
+`pause`: When Cerberus status is 'False', pause Reliability test until Cerberus status is changed to 'True'."
+
+`halt`: When Cerberus status is 'False', halt Reliablity test.
```
It's not super clear here what `halt` does. Does it end the reliability test?

qiliRedHat replied:
Yes, just like touching a halt file. I can make it clearer.

paigerube14 commented:
One more small change that might be nice; adding a logging info statement that we are sleeping for x amount of time in-between runs at this [line](https://github.com/openshift/svt/blob/214a43e7f5c48d062a237c8471989345de750f81/reliability/tasks/TaskManager.py#L209)

paigerube14 commented on reliability/README.md:
````diff
@@ -1,81 +1,205 @@
+
 ```
 
+## Git
````
Might be good to add these sections (Git, Install dependencies, run and control activity execution) to below the files needed section /more toward the top so you don't have to read as much if you are just wanting to know how to install and run

qiliRedHat replied:
Good suggestion from reader's point of view!

mffiedler commented:
@qiliRedHat Go ahead and squash when you are ready and I will do a final review.

qiliRedHat commented:
@mffiedler I squashed the commits. PTAL.
Main changes in this PR:
- Reliability cerberus integration: check Cerberus status and take action accordingly, save history when new failure happens.
- Added a requirements.txt.
- Make log file location configuable #673. 
- Leave only one example config file.
- Update README.md with the recent reliability enhancement #672.

paigerube14 reviewed (approved):
LGTM, great additions!!

mffiedler commented:
Thanks @qiliRedHat !
